File: tools4a-device/src/androidTest/java/com/github/panpf/tools4a/device/test/DevicexTest.java
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.tools4a.device.test;

import android.Manifest;
import android.content.Context;
import android.os.Build;

import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;

import com.github.panpf.tools4a.device.Devicex;
import com.github.panpf.tools4a.permission.Permissionx;
import com.github.panpf.tools4j.lang.Stringx;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.regex.Pattern;

@RunWith(AndroidJUnit4.class)
public class DevicexTest {

    @Test
    public final void testModel() {
        Assert.assertTrue(Stringx.isSafe(Devicex.getModel()));
    }

    @Test
    public final void testName() {
        Assert.assertTrue(Stringx.isSafe(Devicex.getDevice()));
    }

    @Test
    public final void testHardware() {
        Assert.assertTrue(Stringx.isSafe(Devicex.getHardware()));
    }

    @Test
    public final void testProduct() {
        Assert.assertTrue(Stringx.isSafe(Devicex.getProduct()));
    }

    @Test
    public final void testBrand() {
        Assert.assertTrue(Stringx.isSafe(Devicex.getBrand()));
    }

    @Test
    public final void testSupportedAbis() {
        Assert.assertTrue(Devicex.getSupportedAbis().length != 0);
    }

    @Test
    public final void testPhoneNumber() {
        Context context = InstrumentationRegistry.getContext();
        Assert.assertNotNull(Devicex.getPhoneNumberOr(context, "defaultValue"));
    }

    @Test
    public final void testAndroidId() {
        String androidId = Devicex.getAndroidIdOr(InstrumentationRegistry.getContext(), "defaultValue");
        Assert.assertTrue("androidId: " + androidId, Stringx.isSafe(androidId));
    }

    @Test
    public final void testDeviceId() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String deviceId = Devicex.getDeviceIdOr(context, "defaultValue");
            Assert.assertEquals("deviceId: " + deviceId, "defaultValue", deviceId);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String deviceId = Devicex.getDeviceIdOr(context, "defaultValue");
                Assert.assertTrue("deviceId: " + deviceId,
                        Stringx.isSafe(deviceId) && !deviceId.equals("defaultValue")
                );
            } else {
                String deviceId = Devicex.getDeviceIdOr(context, "defaultValue");
                Assert.assertEquals("deviceId: " + deviceId, "defaultValue", deviceId);
            }
        } else {
            String deviceId = Devicex.getDeviceIdOr(context, "defaultValue");
            Assert.assertTrue("deviceId: " + deviceId,
                    Stringx.isSafe(deviceId) && !deviceId.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testSubscriberId() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String subscriberId = Devicex.getSubscriberIdOr(context, "defaultValue");
            Assert.assertEquals("subscriberId: " + subscriberId, "defaultValue", subscriberId);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String subscriberId = Devicex.getSubscriberIdOr(context, "defaultValue");
                Assert.assertTrue("subscriberId: " + subscriberId,
                        Stringx.isSafe(subscriberId) && !subscriberId.equals("defaultValue")
                );
            } else {
                String subscriberId = Devicex.getSubscriberIdOr(context, "defaultValue");
                Assert.assertEquals("subscriberId: " + subscriberId, "defaultValue", subscriberId);
            }
        } else {
            String subscriberId = Devicex.getSubscriberIdOr(context, "defaultValue");
            Assert.assertTrue("subscriberId: " + subscriberId,
                    Stringx.isSafe(subscriberId) && !subscriberId.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testSimSerialNumber() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String simSerialNumber = Devicex.getSimSerialNumberOr(context, "defaultValue");
            Assert.assertEquals("simSerialNumber: " + simSerialNumber, "defaultValue", simSerialNumber);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String simSerialNumber = Devicex.getSimSerialNumberOr(context, "defaultValue");
                Assert.assertTrue("simSerialNumber: " + simSerialNumber,
                        Stringx.isSafe(simSerialNumber) && !simSerialNumber.equals("defaultValue"));
            } else {
                String simSerialNumber = Devicex.getSimSerialNumberOr(context, "defaultValue");
                Assert.assertEquals("simSerialNumber: " + simSerialNumber, "defaultValue", simSerialNumber);
            }
        } else {
            String simSerialNumber = Devicex.getSimSerialNumberOr(context, "defaultValue");
            Assert.assertTrue("simSerialNumber: " + simSerialNumber,
                    Stringx.isSafe(simSerialNumber) && !simSerialNumber.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testSerial() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String serial = Devicex.getSerialOr("defaultValue");
            Assert.assertEquals("serial: " + serial, "defaultValue", serial);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String serial = Devicex.getSerialOr("defaultValue");
                Assert.assertTrue("serial: " + serial,
                        Stringx.isSafe(serial) && !serial.equals("defaultValue")
                );
            } else {
                String serial = Devicex.getSerialOr("defaultValue");
                Assert.assertEquals("serial: " + serial, "defaultValue", serial);
            }
        } else {
            String serial = Devicex.getSerialOr("defaultValue");
            Assert.assertTrue("serial: " + serial,
                    Stringx.isSafe(serial) && !serial.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testIMEI() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String imei = Devicex.getIMEIOr(context, "defaultValue");
            Assert.assertEquals("imei: " + imei, "defaultValue", imei);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String imei = Devicex.getIMEIOr(context, "defaultValue");
                Assert.assertTrue("imei: " + imei,
                        Stringx.isSafe(imei) && !imei.equals("defaultValue")
                );
            } else {
                String imei = Devicex.getIMEIOr(context, "defaultValue");
                Assert.assertEquals("imei: " + imei, "defaultValue", imei);
            }
        } else {
            String imei = Devicex.getIMEIOr(context, "defaultValue");
            Assert.assertTrue("imei: " + imei,
                    Stringx.isSafe(imei) && !imei.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testIMSI() {
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            String imsi = Devicex.getIMSIOr(context, "defaultValue");
            Assert.assertEquals("imsi: " + imsi, "defaultValue", imsi);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.READ_PHONE_STATE)) {
                String imsi = Devicex.getIMSIOr(context, "defaultValue");
                Assert.assertTrue("imsi: " + imsi,
                        Stringx.isSafe(imsi) && !imsi.equals("defaultValue")
                );
            } else {
                String imsi = Devicex.getIMSIOr(context, "defaultValue");
                Assert.assertEquals("imsi: " + imsi, "defaultValue", imsi);
            }
        } else {
            String imsi = Devicex.getIMSIOr(context, "defaultValue");
            Assert.assertTrue("imsi: " + imsi,
                    Stringx.isSafe(imsi) && !imsi.equals("defaultValue")
            );
        }
    }

    @Test
    public final void testMacAddress() {
        Pattern macAddressPattern = Pattern.compile("([A-Fa-f0-9]{2}(-[A-Fa-f0-9]{2}){5})|([A-Fa-f0-9]{2}(:[A-Fa-f0-9]{2}){5})");
        Context context = InstrumentationRegistry.getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Permissionx.isGrantPermissions(context, Manifest.permission.ACCESS_WIFI_STATE)) {
                Assert.assertTrue(macAddressPattern.matcher(Devicex.getMacAddressOr(context, "defaultValue")).matches());
            } else {
                Assert.assertEquals("02:00:00:00:00:00", Devicex.getMacAddress(context));
                Assert.assertEquals("defaultValue", Devicex.getMacAddressOr(context, "defaultValue"));
            }
        } else {
            Assert.assertTrue(macAddressPattern.matcher(Devicex.getMacAddressOr(context, "defaultValue")).matches());
        }
    }
}

File: tools4a-device/src/androidTest/java/com/github/panpf/tools4a/device/test/DevicexTest.kt
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.tools4a.device.test

import android.Manifest
import android.content.Context
import android.os.Build
import android.telephony.TelephonyManager
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.github.panpf.tools4a.device.Devicex
import com.github.panpf.tools4a.permission.ktx.isGrantPermissions
import com.github.panpf.tools4a.systemproperties.SystemPropertiesx
import com.github.panpf.tools4j.lang.ktx.isSafe
import com.github.panpf.tools4j.test.Assertx.assertThrow
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import java.util.regex.Pattern

@RunWith(AndroidJUnit4::class)
class DevicexTest {
    @Test
    fun testIsEmulator() {
        assertEquals("1" == SystemPropertiesx.get("ro.kernel.qemu"), Devicex.isEmulator())
    }

    @Test
    fun testGetProduct() {
        assertTrue(Devicex.getProduct().isSafe())
    }

    @Test
    fun testGetBrand() {
        assertTrue(Devicex.getBrand().isSafe())
    }

    @Test
    fun testGetModel() {
        assertTrue(Devicex.getModel().isSafe())
    }

    @Test
    fun testGetDevice() {
        assertTrue(Devicex.getDevice().isSafe())
    }

    @Test
    fun testGetHardware() {
        assertTrue(Devicex.getHardware().isSafe())
    }

    @Test
    fun testGetSupportedAbis() {
        assertTrue(Devicex.getSupportedAbis().isNotEmpty())
    }

    @Test
    fun testGetPhoneNumber() {
        val context = InstrumentationRegistry.getInstrumentation().context
        var cause: Exception? = null
        val phoneNumber = try {
            val manager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
            manager.line1Number
        } catch (e: Exception) {
            cause = e
            e.printStackTrace()
            null
        }
        if (phoneNumber != null && phoneNumber.isNotEmpty() && !"unknown".equals(phoneNumber, ignoreCase = true)) {
            assertEquals(phoneNumber, Devicex.getPhoneNumberOr(context, "defaultValue"))
            assertEquals(phoneNumber, Devicex.getPhoneNumberOrThrow(context))
            assertEquals(phoneNumber, Devicex.getPhoneNumberOrNull(context))
        } else {
            assertEquals("defaultValue", Devicex.getPhoneNumberOr(context, "defaultValue"))
            if (cause != null && cause is SecurityException) {
                assertThrow(SecurityException::class.java) { Devicex.getPhoneNumberOrThrow(context) }
            } else {
                assertThrow(IllegalStateException::class.java) { Devicex.getPhoneNumberOrThrow(context) }
            }
            assertNull(Devicex.getPhoneNumberOrNull(context))
        }
    }

    @Test
    fun testGetDeviceId() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getDeviceIdOr(context, "defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getDeviceIdOrThrow(context) }
            assertNull(Devicex.getDeviceIdOrNull(context))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getDeviceIdOr(context, "defaultValue") != "defaultValue")
                assertTrue(Devicex.getDeviceIdOrThrow(context).isSafe())
                assertTrue(Devicex.getDeviceIdOrNull(context).isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getDeviceIdOr(context, "defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getDeviceIdOrThrow(context) }
                assertNull(Devicex.getDeviceIdOrNull(context))
            }
        } else {
            assertTrue(Devicex.getDeviceIdOr(context, "defaultValue") != "defaultValue")
            assertTrue(Devicex.getDeviceIdOrThrow(context).isSafe())
            assertTrue(Devicex.getDeviceIdOrNull(context).isSafe())
        }
    }

    @Test
    fun testGetAndroidId() {
        val context = InstrumentationRegistry.getInstrumentation().context
        assertTrue(Devicex.getAndroidIdOr(context, "defaultValue").isSafe())
        assertTrue(Devicex.getAndroidIdOrThrow(context).isSafe())
        assertTrue(Devicex.getAndroidIdOrNull(context).isSafe())
    }

    @Test
    fun testGetSubscriberId() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getSubscriberIdOr(context, "defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getSubscriberIdOrThrow(context) }
            assertNull(Devicex.getSubscriberIdOrNull(context))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getSubscriberIdOr(context, "defaultValue") != "defaultValue")
                assertTrue(Devicex.getSubscriberIdOrThrow(context).isSafe())
                assertTrue(Devicex.getSubscriberIdOrThrow(context).isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getSubscriberIdOr(context, "defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getSubscriberIdOrThrow(context) }
                assertNull(Devicex.getSubscriberIdOrNull(context))
            }
        } else {
            assertTrue(Devicex.getSubscriberIdOr(context, "defaultValue") != "defaultValue")
            assertTrue(Devicex.getSubscriberIdOrThrow(context).isSafe())
            assertTrue(Devicex.getSubscriberIdOrThrow(context).isSafe())
        }
    }

    @Test
    fun testGetSimSerialNumber() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getSimSerialNumberOr(context, "defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getSimSerialNumberOrThrow(context) }
            assertNull(Devicex.getSimSerialNumberOrNull(context))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getSimSerialNumberOr(context, "defaultValue") != "defaultValue")
                assertTrue(Devicex.getSimSerialNumberOrThrow(context).isSafe())
                assertTrue(Devicex.getSimSerialNumberOrNull(context).isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getSimSerialNumberOr(context, "defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getSimSerialNumberOrThrow(context) }
                assertNull(Devicex.getSimSerialNumberOrNull(context))
            }
        } else {
            assertTrue(Devicex.getSimSerialNumberOr(context, "defaultValue") != "defaultValue")
            assertTrue(Devicex.getSimSerialNumberOrThrow(context).isSafe())
            assertTrue(Devicex.getSimSerialNumberOrNull(context).isSafe())
        }
    }

    @Test
    fun testGetSerial() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getSerialOr("defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getSerialOrThrow() }
            assertNull(Devicex.getSerialOrNull())
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getSerialOr("defaultValue") != "defaultValue")
                assertTrue(Devicex.getSerialOrThrow().isSafe())
                assertTrue(Devicex.getSerialOrNull().isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getSerialOr("defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getSerialOrThrow() }
                assertNull(Devicex.getSerialOrNull())
            }
        } else {
            assertTrue(Devicex.getSerialOr("defaultValue") != "defaultValue")
            assertTrue(Devicex.getSerialOrThrow().isSafe())
            assertTrue(Devicex.getSerialOrNull().isSafe())
        }
    }

    @Test
    fun testGetIMEI() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getIMEIOr(context, "defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getIMEIOrThrow(context) }
            assertNull(Devicex.getIMEIOrNull(context))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getIMEIOr(context, "defaultValue") != "defaultValue")
                assertTrue(Devicex.getIMEIOrThrow(context).isSafe())
                assertTrue(Devicex.getIMEIOrNull(context).isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getIMEIOr(context, "defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getIMEIOrThrow(context) }
                assertNull(Devicex.getIMEIOrNull(context))
            }
        } else {
            assertTrue(Devicex.getIMEIOr(context, "defaultValue") != "defaultValue")
            assertTrue(Devicex.getIMEIOrThrow(context).isSafe())
            assertTrue(Devicex.getIMEIOrNull(context).isSafe())
        }
    }

    @Test
    fun testGetIMSI() {
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            assertEquals("defaultValue", Devicex.getIMSIOr(context, "defaultValue"))
            assertThrow(SecurityException::class.java) { Devicex.getIMSIOrThrow(context) }
            assertNull(Devicex.getIMSIOrNull(context))
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.READ_PHONE_STATE)) {
                assertTrue(Devicex.getIMSIOr(context, "defaultValue") != "defaultValue")
                assertTrue(Devicex.getIMSIOrThrow(context).isSafe())
                assertTrue(Devicex.getIMSIOrThrow(context).isSafe())
            } else {
                assertEquals("defaultValue", Devicex.getIMSIOr(context, "defaultValue"))
                assertThrow(SecurityException::class.java) { Devicex.getIMSIOrThrow(context) }
                assertNull(Devicex.getIMSIOrNull(context))
            }
        } else {
            assertTrue(Devicex.getIMSIOr(context, "defaultValue") != "defaultValue")
            assertTrue(Devicex.getIMSIOrThrow(context).isSafe())
            assertTrue(Devicex.getIMSIOrThrow(context).isSafe())
        }
    }

    @Test
    fun testGetMacAddress() {
        val macAddressPattern = Pattern.compile("([A-Fa-f0-9]{2}(-[A-Fa-f0-9]{2}){5})|([A-Fa-f0-9]{2}(:[A-Fa-f0-9]{2}){5})")
        val context = InstrumentationRegistry.getInstrumentation().context
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (context.isGrantPermissions(Manifest.permission.ACCESS_WIFI_STATE)) {
                val maxAddress = Devicex.getMacAddressOr(context, "defaultValue")
                assertTrue("maxAddress: $maxAddress", maxAddress == "defaultValue" || macAddressPattern.matcher(maxAddress).matches())
            } else {
                val maxAddress = Devicex.getMacAddressOr(context, "defaultValue")
                val maxAddressNoDefault = Devicex.getMacAddress(context)
                assertEquals("maxAddressNoDefault: $maxAddressNoDefault", "02:00:00:00:00:00", maxAddressNoDefault)
                assertEquals("maxAddress: $maxAddress", "defaultValue", maxAddress)
            }
        } else {
            val maxAddress = Devicex.getMacAddressOr(context, "defaultValue")
            assertTrue("maxAddress: $maxAddress", macAddressPattern.matcher(maxAddress).matches())
        }
    }
}